### PR TITLE
Make array export dynamic

### PR DESCRIPTION
**Pull request by SeboCode**

### CODE REVIEW

File: src/call_wrapper.ts [hunk 1]
```diff
@@ -23,25 +23,24 @@ class CallWrapper {
...
+    private static readMemory(parameters: any[],
+                              outParameters: any[],
+                              fixedParameters: any[],
+                              readableMemory: Uint32Array) {
 
-                for (let j = 0; j < array.length; j++) {
-                    array[j] = readableMemory[fixedParameters[index] / 4 + j];
-                }
+        for (const outParameter of outParameters) {
+            const parameterIndex = parameters.indexOf(outParameter);
```
Comment (InverseIntegral) [inline src/call_wrapper.ts]: I think this could be renamed to `outParameterIndex`

File: src/call_wrapper.ts [hunk 2]
```diff
@@ -23,25 +23,24 @@ class CallWrapper {
...
+            const array = parameters[parameterIndex];
 
-                index++;
+            for (let j = 0; j < array.length; j++) {
+                array[j] = readableMemory[fixedParameters[parameterIndex] / 4 + j];
```
Comment (InverseIntegral) [inline src/call_wrapper.ts]: Move the static part (`fixedParameters[parameterIndex]`) out of the for loop

File: src/call_wrapper.ts [hunk 3]
```diff
@@ -23,25 +23,24 @@ class CallWrapper {
...
+                              outParameters: any[],
+                              fixedParameters: any[],
+                              readableMemory: Uint32Array) {
 
-                for (let j = 0; j < array.length; j++) {
-                    array[j] = readableMemory[fixedParameters[index] / 4 + j];
-                }
+        for (const outParameter of outParameters) {
+            const parameterIndex = parameters.indexOf(outParameter);
+            const array = parameters[parameterIndex];
```
Comment (InverseIntegral) [inline src/call_wrapper.ts]: This should be changed to `outArray` or something similar

Comment (InverseIntegral) [reply src/call_wrapper.ts]: And the index from above could also be `-1` if it doesn't exist in the `parameters` array. This should be checked and an error should get thrown.

File: test/transpiler/transpiler_array_test.ts [hunk 4]
```diff
@@ -122,23 +122,36 @@ describe('Transpiler', () => {
...
 
             const array1 = [0];
-            wrapper.call(array1);
+            wrapper.setOutParameters(array1).call(array1);
             expect(array1[0]).to.equal(5);
 
             const array2 = [1, 2, 3];
-            wrapper.call(array2);
+            wrapper.setOutParameters(array2).call(array2);
             expect(array2[0]).to.equal(5);
```
Comment (InverseIntegral) [inline test/transpiler/transpiler_array_test.ts]: Use `expect(array2).to.eql([5, 2, 3])` here

File: test/transpiler/transpiler_array_test.ts [hunk 5]
```diff
@@ -122,23 +122,36 @@ describe('Transpiler', () => {
...
+        it('should handle array export with non exportable values', () => {
+            const wrapper = transpiler.transpile('function arrayExport(arr, value) { arr[0] = value; return arr[0]; }');
+            wrapper.setFunctionName('arrayExport');
+
+            const array = [0];
+            wrapper.setOutParameters(array).call(array, 3);
+            expect(array[0]).to.equal(3);
+        });
+
         it('should handle multiple array wrapper', () => {
```
Comment (InverseIntegral) [inline test/transpiler/transpiler_array_test.ts]: What do you mean by wrapper here?

File: test/transpiler/transpiler_array_test.ts [hunk 6]
```diff
@@ -148,15 +161,36 @@ describe('Transpiler', () => {
...
+            const array2 = [52, 53];
+
+            wrapper.setFunctionName('arrayExport')
+                .setOutParameters(array2, array1)
+                .call(array1, array2);
+
+            expect(array1[0]).to.equal(61);
+            expect(array2[0]).to.equal(62);
+            expect(array2[1]).to.equal(53);
+        });
```
Comment (InverseIntegral) [inline test/transpiler/transpiler_array_test.ts]: Add a test case where the out parameter is not passed as a parameter

File: test/transpiler/transpiler_array_test.ts (hunk 5 above)
Comment (SeboCode) [reply test/transpiler/transpiler_array_test.ts]: That accidentally happened because of a search & replace.

File: src/call_wrapper.ts [hunk 7]
```diff
@@ -31,9 +31,10 @@ class CallWrapper {
         for (const outParameter of outParameters) {
             const outParameterIndex = parameters.indexOf(outParameter);
             const outArray = parameters[outParameterIndex];
+            const memoryAddress = fixedParameters[outParameterIndex] / 4;
```
Comment (InverseIntegral) [inline src/call_wrapper.ts]: Maybe rename to `memoryIndex` or `memoryBaseIndex`

File: src/call_wrapper.ts [hunk 8]
```diff
@@ -100,6 +105,8 @@ class CallWrapper {
             const readableMemory = new Uint32Array(exportedMemory.buffer);
 
             CallWrapper.readMemory(parameters, this.outParameters, fixedParameters, readableMemory);
+        } else if (this.outParameters !== undefined) {
+            throw new Error('Output parameters with no memory dependent parameters');
```
Comment (InverseIntegral) [inline src/call_wrapper.ts]: I think this branch is never tested

Comment (InverseIntegral) [reply src/call_wrapper.ts]: And maybe use a nested if to improve redability of this if statement:
```javascript
if (this.outParameters !== undefined) {
    if (hasArrayParameters) {

    } else {

    }
}
```

**Pull request merged**